### PR TITLE
added petProfileModel

### PR DESCRIPTION
**Pull request by konradkop**
What this PR does (required):
Created a model schema for Pet profiles

petProfileModel

### CODE REVIEW

File: server/controllers/petProfile.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  res.status(200).json({ petProfile });
+});
+
+exports.updatePetProfile = asyncHandler(async (req, res, next) => {
+  const values = ({ _id, petName, ownerID, petType, petAge, petStatus } =
+    req.body);
+
+  for (let prop in values) {
+    if (!values[prop]) {
+      delete values[prop];  //it will remove fields who are undefined or null
```
Comment (ezeilo-su) [inline server/controllers/petProfile.js:26]: You don't need this comment as this line is descriptive.

File: server/controllers/petProfile.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+  const petProfile = await PetProfile.findOneAndUpdate(
+    { _id: values._id },
+    values,
+    { new: true, strict: false }
+  );
+
+  res.status(200).json({ petProfile });
+});
+exports.findPetProfilesByOwnerId = asyncHandler(async (req, res, next) => {
```
Comment (ezeilo-su) [inline server/controllers/petProfile.js:38]: You can shorten this controller name you will always find pet profile by id, you can use `findPetProfile`

File: server/controllers/petProfile.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  const petProfile = await petProfile.findById({ ownerID: req.body.ownerID });
```
Comment (ezeilo-su) [inline server/controllers/petProfile.js:39]: Ensure to validate `ownerID` in the request body. If its not supplied, return a bad request status to the client.

Comment (ezeilo-su) [inline server/controllers/petProfile.js:39]: Ensure to validate `ownerID` in the request body. If its not supplied, return a bad request status to the client.

File: server/controllers/petProfile.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  if (!petProfile) {
+    res.status(404);
+    throw new Error("No pet profile found for ID");
+  }
+
+  res.status(200).json(petProfile);
+});
+
+exports.findAllPetProfiles = asyncHandler(async (req, res, next) => {
```
Comment (ezeilo-su) [inline server/controllers/petProfile.js:48]: This cane be `getAllPetProfiles`

File: server/models/PetProfile.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    ownerID: {
+      type: String,
+      required: true,
+    },
+    petType: {
+        type: String,
+    },
+    petAge: {
+        type: Number,
+    },
```
Comment (ezeilo-su) [inline server/models/PetProfile.js:18]: Make sure to use proper indentation

File: server/models/PetProfile.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    petStatus: {
+      type: String,
+      enum: [ "availabile", "not available", null ],
+  },
+  },
+  {
+    timestamps: true,
+  },
+);
```
Comment (ezeilo-su) [inline server/models/PetProfile.js:27]: Ensure to format code properly

File: server/routes/petProfile.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+router.route('/createPetProfile').post(
+    protect, 
+    [check('petName').not().isEmpty(),
+    check('petType').isString(),
+    check('petAge').isNumeric(),
+    check('petStatus').isString(),
+    (req, res, next) => {
+        const errors = validationResult(req);
+    
+        console.log(errors);
```
Comment (ezeilo-su) [inline server/routes/petProfile.js:21]: Remove `console.log`

File: server/routes/petProfile.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+router.route('/updatePetProfile').post(
+    protect,
+    [check('petName').not().isEmpty(),
+    check('petType').isString(),
+    check('petAge').isNumeric(),
+    check('petStatus').isString(),
+    (req, res, next) => {
+        const errors = validationResult(req);
+    
+        console.log(errors);
```
Comment (ezeilo-su) [inline server/routes/petProfile.js:37]: Remove `console.log`

**Review by ezeilo-su — CHANGES_REQUESTED** [review]
Your code looks good. You should always remember to link the ticket in your PR. Also, check the few comments that I left below and address them.
Happy coding!

File: server/controllers/petProfile.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
+const PetProfile = require("../models/PetProfile");
+const asyncHandler = require("express-async-handler");
+
+exports.createPetProfile = asyncHandler(async (req, res, next) => {
+  const { petName, ownerID, petType, petAge, petStatus } = req.body;
+
+
+
```
[the remark this answers is not in the document]
Comment (konradkop) [reply server/controllers/petProfile.js:8]: OK

File: server/controllers/petProfile.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  const petProfile = await PetProfile.create({
```
[the remark this answers is not in the document]
Comment (konradkop) [reply server/controllers/petProfile.js:9]: OK

File: server/controllers/petProfile.js (hunk 1 above)
Comment (konradkop) [reply server/controllers/petProfile.js:26]: OK

File: server/controllers/petProfile.js (hunk 2 above)
Comment (konradkop) [reply server/controllers/petProfile.js:38]: OK

File: server/controllers/petProfile.js (hunk 3 above)
Comment (konradkop) [reply server/controllers/petProfile.js:39]: OK

Comment (konradkop) [reply server/controllers/petProfile.js:39]: OK

File: server/controllers/petProfile.js (hunk 4 above)
Comment (konradkop) [reply server/controllers/petProfile.js:48]: OK

File: server/routes/petProfile.js (hunk 8 above)
Comment (konradkop) [reply server/routes/petProfile.js:37]: OK

File: server/routes/petProfile.js (hunk 7 above)
Comment (konradkop) [reply server/routes/petProfile.js:21]: OK

File: server/models/PetProfile.js (hunk 6 above)
Comment (konradkop) [reply server/models/PetProfile.js:27]: OK

**Comment by konradkop** [comment]
Ok I have addressed your comments. Please let me know if I can merge this PR

File: server/controllers/petProfile.js [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+const asyncHandler = require("express-async-handler");
+
+exports.createPetProfile = asyncHandler(async (req, res, next) => {
+  const { petName, ownerID, petType, petAge, petStatus } = req.body;
+  if (!petName) {
+    res.status(404);
+    throw new Error("No pet name provided");
+  }
+  if (!ownerID) {
+    res.status(404);
```
Comment (moffatethan) [inline server/controllers/petProfile.js:11]: These aren't the right status codes here, the right status code would be 400 HTTP code.

File: server/controllers/petProfile.js [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+exports.updatePetProfile = asyncHandler(async (req, res, next) => {
+  const values = ({ _id, petName, ownerID, petType, petAge, petStatus } =
+    req.body);
+    if (!petName) {
+      res.status(404);
+      throw new Error("No pet name provided");
+    }
+    if (!ownerID) {
+      res.status(404);
```
Comment (moffatethan) [inline server/controllers/petProfile.js:33]: Same here

File: server/controllers/petProfile.js [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+    { _id: values._id },
+    values,
+    { new: true, strict: false }
+  );
+
+  res.status(200).json({ petProfile });
+});
+exports.findPetProfiles = asyncHandler(async (req, res, next) => {
+  if (!ownerID) {
+    res.status(404);
```
Comment (moffatethan) [inline server/controllers/petProfile.js:53]: And here as well

File: server/controllers/petProfile.js [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+    throw new Error("No owner id provided");
+  }
+  const petProfile = await petProfile.findById({ ownerID: req.body.ownerID });
+  if (!petProfile) {
+    res.status(404);
```
Comment (moffatethan) [inline server/controllers/petProfile.js:58]: Right here - everywhere applicable please address!

File: server/controllers/petProfile.js [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+  res.status(200).json(petProfile);
+});
+
+exports.getAllPetProfiles = asyncHandler(async (req, res, next) => {
+  const petProfiles = await PetProfile.find({});
+
+  if (!petProfiles) {
+    res.status(404);
+    throw new Error("No profiles found");
+  }
```
Comment (moffatethan) [inline server/controllers/petProfile.js:71]: This shouldn't be here, if there are not profiles then we can just respond with an empty array

File: server/routes/petProfile.js [hunk 16]
```diff
@@ -0,0 +1,44 @@
+const express = require('express');
+const protect = require("../middleware/auth");
```
Comment (moffatethan) [inline server/routes/petProfile.js:2]: Good job!

File: server/controllers/petProfile.js (hunk 11 above)
Comment (moffatethan) [reply server/controllers/petProfile.js:11]: Q: If we have middleware validation why do we need to do this actually? We could juse verify no errors occured in the validation result as opposed to doing this here?

**Review by moffatethan — CHANGES_REQUESTED** [review]
Left you some comments @konradkop, looks good!

File: server/controllers/petProfile.js (hunk 11 above)
Comment (konradkop) [reply server/controllers/petProfile.js:11]: OK- makes sense I will remove

File: server/controllers/petProfile.js (hunk 12 above)
Comment (konradkop) [reply server/controllers/petProfile.js:33]: OK- makes sense I will remove

File: server/controllers/petProfile.js (hunk 13 above)
Comment (konradkop) [reply server/controllers/petProfile.js:53]: OK- makes sense I will remove

File: server/controllers/petProfile.js (hunk 14 above)
Comment (konradkop) [reply server/controllers/petProfile.js:58]: OK- makes sense I will remove

File: server/controllers/petProfile.js (hunk 15 above)
Comment (konradkop) [reply server/controllers/petProfile.js:71]: OK

**Review by moffatethan — APPROVED** [review]
Good work

**Pull request merged**